lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -95,6 +95,39 @@ bool firrtl::isDuplexValue(Value val) {
       .Default([](auto) { return false; });
 }
 
+/// Return the kind of port this is given the port type from a 'mem' decl.
+static MemOp::PortKind getMemPortKindFromType(FIRRTLType type) {
+  // Get the kind of port based on the fields of the Bundle.
+  auto portType = type.dyn_cast<BundleType>();
+  if (!portType)
+    return MemOp::PortKind::Debug;
+  unsigned fields = 0;
+  // Get the kind of port based on the fields of the Bundle.
+  for (auto elem : portType.getElements()) {
+    fields |= llvm::StringSwitch<unsigned>(elem.name.getValue())
+                  .Case("addr", 1)
+                  .Case("en", 2)
+                  .Case("clk", 4)
+                  .Case("data", 8)
+                  .Case("mask", 16)
+                  .Case("rdata", 32)
+                  .Case("wdata", 64)
+                  .Case("wmask", 128)
+                  .Case("wmode", 256)
+                  .Default(512);
+  }
+  // addr, en, clk, data
+  if (fields == 15)
+    return MemOp::PortKind::Read;
+  // addr, en, clk, data, mask
+  if (fields == 31)
+    return MemOp::PortKind::Write;
+  // addr, en, clk, wdata, wmask, rdata, wmode
+  if (fields == 487)
+    return MemOp::PortKind::ReadWrite;
+  return MemOp::PortKind::Debug;
+}
```

Comment:
I would be tempted to write this using actual numbers or'd together in order to avoid the magical constants and comments. (Let the compiler statically figure them out.) Something like:

```diff
diff --git a/lib/Dialect/FIRRTL/FIRRTLOps.cpp b/lib/Dialect/FIRRTL/FIRRTLOps.cpp
index f36b3fdc8..0a809c3c0 100644
--- a/lib/Dialect/FIRRTL/FIRRTLOps.cpp
+++ b/lib/Dialect/FIRRTL/FIRRTLOps.cpp
@@ -97,6 +97,16 @@ bool firrtl::isDuplexValue(Value val) {
 
 /// Return the kind of port this is given the port type from a 'mem' decl.
 static MemOp::PortKind getMemPortKindFromType(FIRRTLType type) {
+  constexpr unsigned int addr = 1;
+  constexpr unsigned int en = 2;
+  constexpr unsigned int clk = 4;
+  constexpr unsigned int data = 8;
+  constexpr unsigned int mask = 16;
+  constexpr unsigned int rdata = 32;
+  constexpr unsigned int wdata = 64;
+  constexpr unsigned int wmask = 128;
+  constexpr unsigned int wmode = 256;
+
   // Get the kind of port based on the fields of the Bundle.
   auto portType = type.dyn_cast<BundleType>();
   if (!portType)
@@ -105,25 +115,22 @@ static MemOp::PortKind getMemPortKindFromType(FIRRTLType type) {
   // Get the kind of port based on the fields of the Bundle.
   for (auto elem : portType.getElements()) {
     fields |= llvm::StringSwitch<unsigned>(elem.name.getValue())
-                  .Case("addr", 1)
-                  .Case("en", 2)
-                  .Case("clk", 4)
-                  .Case("data", 8)
-                  .Case("mask", 16)
-                  .Case("rdata", 32)
-                  .Case("wdata", 64)
-                  .Case("wmask", 128)
-                  .Case("wmode", 256)
+                  .Case("addr", addr)
+                  .Case("en", en)
+                  .Case("clk", clk)
+                  .Case("data", data)
+                  .Case("mask", mask)
+                  .Case("rdata", rdata)
+                  .Case("wdata", wdata)
+                  .Case("wmask", wmask)
+                  .Case("wmode", wmode)
                   .Default(512);
   }
-  // addr, en, clk, data
-  if (fields == 15)
+  if (fields == (addr | en | clk | data))
     return MemOp::PortKind::Read;
-  // addr, en, clk, data, mask
-  if (fields == 31)
+  if (fields == (addr | en | clk | data | mask))
     return MemOp::PortKind::Write;
-  // addr, en, clk, wdata, wmask, rdata, wmode
-  if (fields == 487)
+  if (fields == (addr | en | clk | wdata | wmask | rdata | wmode))
     return MemOp::PortKind::ReadWrite;
   return MemOp::PortKind::Debug;
 }
```

I would be further tempted to write them as `#define`, but this is really showing my C-code thinking...

Reply:
This is a Nice idea, the magic constants were bothering me too.